### PR TITLE
Cluster cache

### PR DESCRIPTION
**Pull request by LorenzoCianciaruso**

### CODE REVIEW

File: pipe-http-server-cloud/src/main/java/com/tesco/aqueduct/pipe/http/Bindings.java [hunk 1]
```diff
@@ -46,6 +50,15 @@ PostgresqlStorage bindPostgreSQL(
         );
     }
 
+    @Singleton
+    LocationResolver locationResolver(
+        @Named("pipe") final DataSource dataSource,
+        @Value("${location.cluster-cache.expire-after-write:1h}") final Duration duration,
```
Comment (adam-yorwerth) [inline pipe-http-server-cloud/src/main/java/com/tesco/aqueduct/pipe/http/Bindings.java:56]: `duration` feels a bit generic, can we do `clusterCacheExpiry` perhaps?

File: pipe-storage-postgresql/src/main/java/com/tesco/aqueduct/pipe/storage/ClusterStorage.java [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+    }
+
+    private List<Long> resolveClusterIdsFor(List<String> resolvedClusterUuids, Connection newConnection) {
+        insertClusterUuids(resolvedClusterUuids, newConnection);
+        return fetchClusterIdsFor(resolvedClusterUuids, newConnection);
+    }
+
+    private Optional<ClusterCache> getClusterIdsFromCache(String locationUuid) {
+        long start = System.currentTimeMillis();
+        try (Connection connection = dataSource.getConnection()) {
```
Comment (adam-yorwerth) [inline pipe-storage-postgresql/src/main/java/com/tesco/aqueduct/pipe/storage/ClusterStorage.java:99]: Getting a new connection here shouldn't be required - I think we really need to have the same connection throughout a request when we're on the happy path.

**Review by adam-yorwerth — COMMENTED** [review]
We need to review how efficient getting 2 connections per request is - I've got concerns about this.

File: pipe-storage-postgresql/src/main/java/com/tesco/aqueduct/pipe/storage/ClusterStorage.java (hunk 2 above)
Comment (LorenzoCianciaruso) [reply pipe-storage-postgresql/src/main/java/com/tesco/aqueduct/pipe/storage/ClusterStorage.java:99]: I would prefer to have 2 different connection if as it is it doesn't affect the performance, having 2 different connections makes the code cleaner and simpler. It performance are affected than we can look into optimise the connections

File: pipe-http-server-cloud/src/integration/groovy/LocationServiceClientIntegrationSpec.groovy [hunk 3]
```diff
@@ -60,11 +60,11 @@ class LocationServiceClientIntegrationSpec extends Specification {
                 .properties(
                     parseYamlConfig(
                     """
-                    micronaut.caches.cluster-cache..expire-after-write: $CACHE_EXPIRY_HOURS
                     location:
                         url:                    $locationBasePath
                         clusters.get.path:      ${LOCATION_CLUSTER_PATH+"?"+CLUSTER_QUERY_PARAM+"="+CLUSTER_QUERY_PARAM_VALUE}
                         clusters.get.path.filter.pattern: $LOCATION_CLUSTER_PATH_FILTER_PATTERN
+                        clusters.cache.expire-after-write: $CACHE_EXPIRY_HOURS
```
Comment (sebastianlavender) [inline pipe-http-server-cloud/src/integration/groovy/LocationServiceClientIntegrationSpec.groovy:67]: Not sure this needs to be extracted out as it's only used in this one place and not checked against in any test. Would make the test a bit cleaner

File: pipe-http-server-cloud/src/integration/groovy/LocationRoutingIntegrationSpec.groovy [hunk 4]
```diff
@@ -252,7 +207,7 @@ class LocationRoutingIntegrationSpec extends Specification {
             .statusCode(200)
 
         and: "response body has messages only for the given location"
```
Comment (sebastianlavender) [inline pipe-http-server-cloud/src/integration/groovy/LocationRoutingIntegrationSpec.groovy:209]: I this step could be renamed to something like:

"response body has no messages for the given location"

File: pipe-http-server-cloud/src/main/java/com/tesco/aqueduct/pipe/http/Bindings.java [hunk 5]
```diff
@@ -46,6 +50,15 @@ PostgresqlStorage bindPostgreSQL(
         );
     }
 
+    @Singleton
+    LocationResolver locationResolver(
+        @Named("pipe") final DataSource dataSource,
+        @Value("${location.clusters.cache.expire-after-write:1h}") final Duration expireAfter,
```
Comment (sebastianlavender) [inline pipe-http-server-cloud/src/main/java/com/tesco/aqueduct/pipe/http/Bindings.java:56]: Why do we default the value here? Isn't that counter intuitive to having the value in the application yml

**Review by sebastianlavender — APPROVED** [review]
LGTM

**Review by sebastianlavender — APPROVED** [review]

**Pull request merged**